src/member.rs

```diff
@@ -0,0 +1,478 @@
+        self.our_keys.node_id().cmp(&other.our_keys.public_id)
+    }
+}
+
+impl PartialOrd for Inner {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        self.our_keys
+            .public_id()
+            .partial_cmp(other.our_keys.public_id())
+    }
+}
+
+impl Eq for Inner {}
+
+impl Drop for Inner {
+    fn drop(&mut self) {
+        self.terminate();
+        thread::sleep(Duration::from_millis(50));
+    }
+}
+
+impl Inner {
+    // Connects to every Node in the group
+    pub fn connect_to_all(&mut self) {
+        for member in self.group.iter() {
+            let (socket_addr, peer) = member.1;
+            // Safe to unwrap
+            let connected = self.connected.get(peer).unwrap();
+            if !connected {
+                self.quic_p2p.connect_to(*socket_addr)
+            }
+        }
+    }
+
+    // Updates the connected list if successfully connected at the transport layer
+    pub fn handle_connected_to(&mut self, peer: Peer) {
+        trace!("Connected to Peer: {:?}", peer);
+        let _ = self.connected.insert(peer, true);
+    }
+
+    // Checks if we are connected to all the nodes in the group
+    pub fn if_connected_to_all(&self) -> bool {
+        self.connected.iter().all(|item| item.1 == &true)
+    }
+
+    // Dispatches the incoming DKG message to the Key_Gen instance
+    fn handle_incoming(&mut self, msg: Bytes, _peer: Peer) {
+        match deserialize(&msg) {
+            Ok(msg) => {
+                let mut rng = thread_rng();
+                if let Some(ref mut key_gen) = self.key_gen {
+                    match key_gen.handle_message(&mut rng, msg) {
+                        Ok(list) => {
+                            for message in list {
+                                // Broadcast the reply messages right away
+                                let _ = self.broadcast(message);
+                            }
+                        }
+                        Err(e) => trace!("Error: {:#?}", e),
+                    }
+                } else {
+                    trace!("Error: No Keygen instance initiated")
+                }
+            }
+            Err(e) => trace!("Error: {:#?}", e),
+        }
+    }
+
+    fn broadcast(&mut self, message: Message<NodeID>) -> Result<(), Error> {
+        for (_, (_socket_addr, peer)) in self.group.iter() {
+            let token = rand::thread_rng().gen();
+            let serialized_msg =
+                serialize(&message).map_err(|e| Error::Serialization(e.to_string()))?;
+            let msg = Bytes::from(serialized_msg.as_slice());
+            self.quic_p2p.send(peer.clone(), msg.clone(), token)
+        }
+        Ok(())
+    }
+
+    // Timed Phase transition needs to be called if we do not finalize automatically
+    fn start_timed_phase_trasition(&mut self) -> Result<Vec<Message<NodeID>>, Error> {
+        let tick = periodic_ms(10_000); // 10 seconds
+
+        tick.recv().unwrap();
```

Comment:
does this blocks the thread for 10 seconds?

Reply:
Yep it does, but it doesn't block the quic-p2p event loop where the messages are being handled. But anyhow, this should be put into a separate thread and make it fire into the main thread.

Reply:
> this should be put into a separate thread and make it fire into the main thread.

+1